### ISSUE TITLE
"IndexError: list index out of range" when accessing a list member that has been appended to

### ISSUE THREAD

**Issue by jckegelman** [nelfin/pylint-protobuf#7]
Hello, I found that `pylint-protobuf` raises an `IndexError` exception when a list member variable is accessed after it has been appended to.

It can be reproduced by running `pylint` with the `pylint-protobuf` plugin on the following code:
```python
class Foo:
    def __init__(self):
        self.bar = []

    def append_bar(self):
        for i in range(3):
            self.bar.append(i)

def main():
    foo = Foo()
    foo.append_bar()
    baz = foo.bar[0]
```

This raises the following exception:
```
Traceback (most recent call last):
  File ".../pylint/__main__.py", line 7, in <module>
    pylint.run_pylint()
  File ".../pylint/__init__.py", line 20, in run_pylint
    Run(sys.argv[1:])
  File ".../pylint/lint.py", line 1628, in __init__
    linter.check(args)
  File ".../pylint/lint.py", line 943, in check
    self._do_check(files_or_modules)
  File ".../pylint/lint.py", line 1075, in _do_check
    self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
  File ".../pylint/lint.py", line 1158, in check_astroid_module
    walker.walk(ast_node)
  File ".../pylint/utils.py", line 1303, in walk
    self.walk(child)
  File ".../pylint/utils.py", line 1303, in walk
    self.walk(child)
  File ".../pylint/utils.py", line 1300, in walk
    cb(astroid)
  File ".../pylint_protobuf/__init__.py", line 398, in visit_assign
    self._visit_assign(node)
  File ".../pylint_protobuf/__init__.py", line 401, in _visit_assign
    new_scope, messages = visit_assign_node(self._scope, self._type_fields, node)
  File ".../pylint_protobuf/__init__.py", line 192, in visit_assign_node
    new_scope.update(_assign(old_scope, target, value))
  File ".../pylint_protobuf/__init__.py", line 149, in _assign
    new_type = _typeof(new_scope, rhs)
  File ".../pylint_protobuf/__init__.py", line 116, in _typeof
    return _typeof(scope, _slice(node))
  File ".../pylint_protobuf/__init__.py", line 67, in _slice
    return mapping[i]
IndexError: list index out of range
```

One idea that seems to address the issue is to add `IndexError` to the list of caught exceptions:
https://github.com/nelfin/pylint-protobuf/blob/b29b5cc26a9a3202e37c4bc369b35f071561197e/pylint_protobuf/__init__.py#L65-L68

thoughts?

**Comment by nelfin** [comment]
You're on a roll @jckegelman, thanks for all of these. It looks like attempting any lookup could cause IndexError to be raised so we should catch it:
```
foo = []
bar = foo[0]
```
also triggers an `IndexError` in slice.

**Comment by jckegelman** [comment]
great, thanks!

I found another one, although dict related instead of list:
```python
foo = None
bar = None
baz = foo[bar]
```
Running `pylint` raises the following exception:
```
Traceback (most recent call last):
...
  File ".../pylint/utils.py", line 1300, in walk
    cb(astroid)
  File ".../pylint_protobuf/__init__.py", line 439, in visit_assign
    self._visit_assign(node)
  File ".../pylint_protobuf/__init__.py", line 442, in _visit_assign
    new_scope, messages = visit_assign_node(self._scope, self._type_fields, node)
  File ".../pylint_protobuf/__init__.py", line 190, in visit_assign_node
    new_scope.update(_assign(old_scope, target, value))
  File ".../pylint_protobuf/__init__.py", line 147, in _assign
    new_type = _typeof(new_scope, rhs)
  File ".../pylint_protobuf/__init__.py", line 114, in _typeof
    return _typeof(scope, _slice(node))
  File ".../pylint_protobuf/__init__.py", line 58, in _slice
    next(k.infer()).value: v for k, v in indexable.items
TypeError: 'method' object is not iterable
```

looks like it's coming from this chunk of code:
https://github.com/nelfin/pylint-protobuf/blob/08703726b9d7b06463c16b73d7eaf76167ab2133/pylint_protobuf/__init__.py#L54-L64

In this case `indexable` is a `NoneType` and has an attribute `items`, but it is a method that returns "the names of locals and their associated node."
http://pylint.pycqa.org/projects/astroid/en/latest/api/base_nodes.html#astroid.scoped_nodes.LocalsDictNodeNG.items

I'm not sure of what should be done instead, either call the method, handle the exception, or something else.

Thanks for all your help!

**Comment by nelfin** [comment]
I think it should just swallow the exception, just because the code would raise an error at runtime shouldn't have any effect on us raising an exception at check-time: that's piercing the veil. It's clear I need some stronger testing/protections around this piece of code given how many things you've found with it recently.

**Comment by nelfin** [comment]
I hope this is sorted for now. Thanks for your help @jckegelman. I'll be releasing these fixes as 0.4.

**Comment by jckegelman** [comment]
0.4 is working great so far, thanks!
